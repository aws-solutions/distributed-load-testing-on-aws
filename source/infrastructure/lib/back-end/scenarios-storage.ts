// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { RemovalPolicy } from "aws-cdk-lib";
import { AttributeType, BillingMode, Table, TableEncryption } from "aws-cdk-lib/aws-dynamodb";
import { AnyPrincipal, Effect, Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { BlockPublicAccess, Bucket, BucketEncryption, HttpMethods } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

export interface ScenarioTestRunnerStorageConstructProps {
  // S3 Logs Bucket
  readonly s3LogsBucket: Bucket;
  // CloudFront domain name
  readonly cloudFrontDomainName: string;
  // Solution Id
  readonly solutionId: string;
}

/**
 * Distributed Load Testing storage construct
 * Creates an S3 bucket to store test scenarios and
 * a Dynamodb table to store tests and test configuration
 */
export class ScenarioTestRunnerStorageConstruct extends Construct {
  public scenariosBucket: Bucket;
  public scenariosS3Policy: Policy;
  public scenariosTable: Table;
  public scenarioDynamoDbPolicy: Policy;
  public historyTable: Table;
  public historyDynamoDbPolicy: Policy;

  constructor(scope: Construct, id: string, props: ScenarioTestRunnerStorageConstructProps) {
    super(scope, id);

    this.scenariosBucket = new Bucket(this, "DLTScenariosBucket", {
      removalPolicy: RemovalPolicy.RETAIN,
      serverAccessLogsBucket: props.s3LogsBucket,
      serverAccessLogsPrefix: "scenarios-bucket-access/",
      encryption: BucketEncryption.KMS_MANAGED,
      enforceSSL: true,
      versioned: true,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      cors: [
        {
          allowedMethods: [HttpMethods.GET, HttpMethods.POST, HttpMethods.PUT],
          allowedOrigins: [`https://${props.cloudFrontDomainName}`],
          allowedHeaders: ["*"],
          exposedHeaders: ["ETag"],
        },
      ],
    });

    this.scenariosBucket.addToResourcePolicy(
      new PolicyStatement({
        actions: ["s3:*"],
        resources: [this.scenariosBucket.bucketArn, `${this.scenariosBucket.bucketArn}/*`],
        effect: Effect.DENY,
        principals: [new AnyPrincipal()],
        conditions: {
          Bool: {
            "aws:SecureTransport": false,
          },
        },
      })
    );

    this.scenariosS3Policy = new Policy(this, "ScenariosS3Policy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["s3:HeadObject", "s3:PutObject", "s3:GetObject", "s3:DeleteObject", "s3:ListBucket"],
          resources: [this.scenariosBucket.bucketArn, `${this.scenariosBucket.bucketArn}/*`],
        }),
      ],
    });

    this.scenariosTable = new Table(this, "DLTScenariosTable", {
      billingMode: BillingMode.PAY_PER_REQUEST,
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: { name: "testId", type: AttributeType.STRING },
      pointInTimeRecovery: true,
    });

    this.historyTable = new Table(this, "DLTHistoryTable", {
      billingMode: BillingMode.PAY_PER_REQUEST,
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: { name: "testId", type: AttributeType.STRING },
      sortKey: { name: "testRunId", type: AttributeType.STRING },
      pointInTimeRecovery: true,
    });

    const historyDDBActions = ["dynamodb:BatchWriteItem", "dynamodb:PutItem", "dynamodb:Query"];
    this.historyDynamoDbPolicy = new Policy(this, "HistoryDynamoDbPolicy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: historyDDBActions,
          resources: [this.historyTable.tableArn],
        }),
      ],
    });

    const scenariosDDBActions = [
      "dynamodb:DeleteItem",
      "dynamodb:GetItem",
      "dynamodb:PutItem",
      "dynamodb:Scan",
      "dynamodb:UpdateItem",
    ];
    this.scenarioDynamoDbPolicy = new Policy(this, "ScenarioDynamoDbPolicy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: scenariosDDBActions,
          resources: [this.scenariosTable.tableArn],
        }),
      ],
    });
  }
}
