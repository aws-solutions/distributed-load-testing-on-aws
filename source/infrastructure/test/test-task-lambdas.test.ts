// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Template } from "aws-cdk-lib/assertions";
import { App, DefaultStackSynthesizer, Stack } from "aws-cdk-lib";
import { TestRunnerLambdasConstruct } from "../lib/back-end/test-task-lambdas";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Effect, Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { LogGroup } from "aws-cdk-lib/aws-logs";
import { AttributeType, Table } from "aws-cdk-lib/aws-dynamodb";

test("DLT Task Lambda Test", () => {
  const app = new App();
  const stack = new Stack(app, "DLTStack", {
    synthesizer: new DefaultStackSynthesizer({
      generateBootstrapVersionRule: false,
    }),
  });

  const testDBPolicy = new Policy(stack, "TestDynamoDBPolicy", {
    statements: [
      new PolicyStatement({
        effect: Effect.DENY,
        actions: ["dynamodb:*"],
        resources: ["*"],
      }),
    ],
  });

  const testS3Policy = new Policy(stack, "TestS3Policy", {
    statements: [
      new PolicyStatement({
        effect: Effect.DENY,
        actions: ["s3:*"],
        resources: ["*"],
      }),
    ],
  });

  const testPolicy = new Policy(stack, "TestPolicy", {
    statements: [
      new PolicyStatement({
        resources: ["*"],
        actions: ["cloudwatch:Get*"],
      }),
    ],
  });

  const testLogGroup = new LogGroup(stack, "TestLogsGroup");

  const testBucket = Bucket.fromBucketName(stack, "SourceCodeBucket", "testbucket");
  const testTable = new Table(stack, "TestTable", {
    partitionKey: {
      name: "id",
      type: AttributeType.STRING,
    },
  });

  const testFunctions = new TestRunnerLambdasConstruct(stack, "TaskRunnerLambdaFunctions", {
    cloudWatchLogsPolicy: testPolicy,
    scenariosDynamoDbPolicy: testDBPolicy,
    ecsTaskExecutionRoleArn: "arn:aws:iam:us-east-1:111122223333:roleArn",
    ecsCloudWatchLogGroup: testLogGroup,
    ecsCluster: "testCluster",
    ecsTaskDefinition: "testTaskDefinition",
    ecsTaskSecurityGroup: "testSecurityGroup",
    historyTable: testTable,
    historyDynamoDbPolicy: testDBPolicy,
    scenariosS3Policy: testS3Policy,
    subnetA: "testSubnetA",
    subnetB: "testSubnetB",
    metricsUrl: "test.example.net",
    sendAnonymizedUsage: "No",
    solutionId: "testId",
    solutionVersion: "testVersion",
    sourceCodeBucket: testBucket,
    sourceCodePrefix: "testPrefix",
    scenariosBucket: "testBucket",
    scenariosTable: testTable,
    uuid: "testId",
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();

  Template.fromStack(stack).hasResourceProperties("AWS::IAM::Policy", {
    PolicyDocument: testPolicy.document.toJSON(),
    Roles: [
      {
        Ref: "TaskRunnerLambdaFunctionsLambdaResultsRole1AF5AB18",
      },
      {
        Ref: "TaskRunnerLambdaFunctionsDLTTestLambdaTaskRoleCB13DE78",
      },
      {
        Ref: "TaskRunnerLambdaFunctionsLambdaTaskCancelerRoleEEC6795B",
      },
      {
        Ref: "TaskRunnerLambdaFunctionsTaskStatusRole4B498DE5",
      },
    ],
  });

  Template.fromStack(stack).hasResourceProperties("AWS::Lambda::Function", {
    Code: {
      S3Bucket: "testbucket",
      S3Key: "testPrefix/results-parser.zip",
    },
    Environment: {
      Variables: {
        METRIC_URL: "test.example.net",
        SCENARIOS_BUCKET: "testBucket",
        SCENARIOS_TABLE: {
          Ref: "TestTable5769773A",
        },
        SEND_METRIC: "No",
        SOLUTION_ID: "testId",
        UUID: "testId",
        VERSION: "testVersion",
      },
    },
    Handler: "index.handler",
    Role: {
      "Fn::GetAtt": ["TaskRunnerLambdaFunctionsLambdaResultsRole1AF5AB18", "Arn"],
    },
    Runtime: "nodejs16.x",
    Timeout: 120,
  });

  Template.fromStack(stack).hasResourceProperties("AWS::IAM::Policy", {
    PolicyDocument: testDBPolicy.document.toJSON(),
    Roles: [
      {
        Ref: "TaskRunnerLambdaFunctionsLambdaResultsRole1AF5AB18",
      },
      {
        Ref: "TaskRunnerLambdaFunctionsDLTTestLambdaTaskRoleCB13DE78",
      },
      {
        Ref: "TaskRunnerLambdaFunctionsTaskStatusRole4B498DE5",
      },
    ],
  });

  Template.fromStack(stack).hasResourceProperties("AWS::IAM::Policy", {
    PolicyDocument: testS3Policy.document.toJSON(),
    Roles: [
      {
        Ref: "TaskRunnerLambdaFunctionsLambdaResultsRole1AF5AB18",
      },
    ],
  });

  Template.fromStack(stack).hasResourceProperties("AWS::IAM::Role", {
    Policies: [
      {
        PolicyDocument: {
          Statement: [
            {
              Action: "ecs:ListTasks",
              Effect: "Allow",
              Resource: "*",
            },
            {
              Action: "ecs:DescribeTasks",
              Effect: "Allow",
              Resource: {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      Ref: "AWS::Partition",
                    },
                    ":ecs:",
                    {
                      Ref: "AWS::Region",
                    },
                    ":",
                    {
                      Ref: "AWS::AccountId",
                    },
                    ":task/*",
                  ],
                ],
              },
            },
          ],
          Version: "2012-10-17",
        },
        PolicyName: "TaskStatusPolicy",
      },
    ],
  });

  expect(testFunctions.resultsParser).toBeDefined();
  expect(testFunctions.taskRunner).toBeDefined();
  expect(testFunctions.taskCanceler).toBeDefined();
  expect(testFunctions.taskCancelerInvokePolicy).toBeDefined();
  expect(testFunctions.taskStatusChecker).toBeDefined();
});
