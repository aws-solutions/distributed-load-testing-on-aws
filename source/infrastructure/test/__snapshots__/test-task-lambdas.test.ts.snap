// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT Task Lambda Test 1`] = `
{
  "Resources": {
    "TaskRunnerLambdaFunctionsDLTTestLambdaTaskRoleCB13DE78": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W11",
              "reason": "ecs:ListTasks does not support resource level permissions",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "ecs:ListTasks",
                  "Effect": "Allow",
                  "Resource": "*",
                },
                {
                  "Action": [
                    "ecs:RunTask",
                    "ecs:DescribeTasks",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":ecs:",
                          {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          {
                            "Ref": "AWS::AccountId",
                          },
                          ":task/*",
                        ],
                      ],
                    },
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":ecs:",
                          {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          {
                            "Ref": "AWS::AccountId",
                          },
                          ":task-definition/*:*",
                        ],
                      ],
                    },
                  ],
                },
                {
                  "Action": "iam:PassRole",
                  "Effect": "Allow",
                  "Resource": "arn:aws:iam:us-east-1:111122223333:roleArn",
                },
                {
                  "Action": "logs:PutMetricFilter",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "TestLogsGroup54B681C7",
                      "Arn",
                    ],
                  },
                },
                {
                  "Action": "cloudwatch:PutDashboard",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":cloudwatch::",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":dashboard/EcsLoadTesting*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "TaskLambdaPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TaskRunnerLambdaFunctionsLambdaResultsPolicy6D0AF2FB": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W12",
              "reason": "The action does not support resource level permissions.",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "cloudwatch:GetMetricWidgetImage",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "logs:DeleteMetricFilter",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TestLogsGroup54B681C7",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TaskRunnerLambdaFunctionsLambdaResultsPolicy6D0AF2FB",
        "Roles": [
          {
            "Ref": "TaskRunnerLambdaFunctionsLambdaResultsRole1AF5AB18",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TaskRunnerLambdaFunctionsLambdaResultsRole1AF5AB18": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W12",
              "reason": "The action does not support resource level permissions.",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TaskRunnerLambdaFunctionsLambdaTaskCancelerRoleEEC6795B": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W11",
              "reason": "ecs:ListTasks does not support resource level permissions",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "ecs:ListTasks",
                  "Effect": "Allow",
                  "Resource": "*",
                },
                {
                  "Action": "ecs:StopTask",
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":ecs:",
                          {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          {
                            "Ref": "AWS::AccountId",
                          },
                          ":task/*",
                        ],
                      ],
                    },
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":ecs:",
                          {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          {
                            "Ref": "AWS::AccountId",
                          },
                          ":task-definition/*:*",
                        ],
                      ],
                    },
                  ],
                },
                {
                  "Action": "dynamodb:UpdateItem",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "TestTable5769773A",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "TaskCancelerPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TaskRunnerLambdaFunctionsResultsParserD7D42F22": {
      "DependsOn": [
        "TaskRunnerLambdaFunctionsLambdaResultsRole1AF5AB18",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
            },
            {
              "id": "W89",
              "reason": "This Lambda function does not require a VPC",
            },
            {
              "id": "W92",
              "reason": "Does not run concurrent executions",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": "testbucket",
          "S3Key": "testPrefix/results-parser.zip",
        },
        "Description": "Result parser for indexing xml test results to DynamoDB",
        "Environment": {
          "Variables": {
            "HISTORY_TABLE": {
              "Ref": "TestTable5769773A",
            },
            "METRIC_URL": "test.example.net",
            "SCENARIOS_BUCKET": "testBucket",
            "SCENARIOS_TABLE": {
              "Ref": "TestTable5769773A",
            },
            "SEND_METRIC": "No",
            "SOLUTION_ID": "testId",
            "UUID": "testId",
            "VERSION": "testVersion",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TaskRunnerLambdaFunctionsLambdaResultsRole1AF5AB18",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TaskRunnerLambdaFunctionsTaskCancelerDAA76ED9": {
      "DependsOn": [
        "TaskRunnerLambdaFunctionsLambdaTaskCancelerRoleEEC6795B",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
            },
            {
              "id": "W89",
              "reason": "This Lambda function does not require a VPC",
            },
            {
              "id": "W92",
              "reason": "Does not run concurrent executions",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": "testbucket",
          "S3Key": "testPrefix/task-canceler.zip",
        },
        "Description": "Stops ECS task",
        "Environment": {
          "Variables": {
            "METRIC_URL": "test.example.net",
            "SCENARIOS_TABLE": {
              "Ref": "TestTable5769773A",
            },
            "SOLUTION_ID": "testId",
            "VERSION": "testVersion",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TaskRunnerLambdaFunctionsLambdaTaskCancelerRoleEEC6795B",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TaskRunnerLambdaFunctionsTaskCancelerInvokePolicyD51068E2": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TaskRunnerLambdaFunctionsTaskCancelerDAA76ED9",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TaskRunnerLambdaFunctionsTaskCancelerInvokePolicyD51068E2",
        "Roles": [
          {
            "Ref": "TaskRunnerLambdaFunctionsTaskStatusRole4B498DE5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TaskRunnerLambdaFunctionsTaskRunner626792CE": {
      "DependsOn": [
        "TaskRunnerLambdaFunctionsDLTTestLambdaTaskRoleCB13DE78",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
            },
            {
              "id": "W89",
              "reason": "This Lambda function does not require a VPC",
            },
            {
              "id": "W92",
              "reason": "Does not run concurrent executions",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": "testbucket",
          "S3Key": "testPrefix/task-runner.zip",
        },
        "Description": "Task runner for ECS task definitions",
        "Environment": {
          "Variables": {
            "SCENARIOS_BUCKET": "testBucket",
            "SCENARIOS_TABLE": {
              "Ref": "TestTable5769773A",
            },
            "SOLUTION_ID": "testId",
            "VERSION": "testVersion",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TaskRunnerLambdaFunctionsDLTTestLambdaTaskRoleCB13DE78",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TaskRunnerLambdaFunctionsTaskStatusCheckerBA69E13B": {
      "DependsOn": [
        "TaskRunnerLambdaFunctionsTaskStatusRole4B498DE5",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
            },
            {
              "id": "W89",
              "reason": "This Lambda function does not require a VPC",
            },
            {
              "id": "W92",
              "reason": "Does not run concurrent executions",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": "testbucket",
          "S3Key": "testPrefix/task-status-checker.zip",
        },
        "Description": "Task status checker",
        "Environment": {
          "Variables": {
            "SCENARIOS_TABLE": {
              "Ref": "TestTable5769773A",
            },
            "SOLUTION_ID": "testId",
            "TASK_CANCELER_ARN": {
              "Fn::GetAtt": [
                "TaskRunnerLambdaFunctionsTaskCancelerDAA76ED9",
                "Arn",
              ],
            },
            "VERSION": "testVersion",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TaskRunnerLambdaFunctionsTaskStatusRole4B498DE5",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 180,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TaskRunnerLambdaFunctionsTaskStatusRole4B498DE5": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W11",
              "reason": "ecs:ListTasks does not support resource level permissions",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "ecs:ListTasks",
                  "Effect": "Allow",
                  "Resource": "*",
                },
                {
                  "Action": "ecs:DescribeTasks",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":ecs:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":task/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "TaskStatusPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestDynamoDBPolicy7AA7B6CD": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:*",
              "Effect": "Deny",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestDynamoDBPolicy7AA7B6CD",
        "Roles": [
          {
            "Ref": "TaskRunnerLambdaFunctionsLambdaResultsRole1AF5AB18",
          },
          {
            "Ref": "TaskRunnerLambdaFunctionsDLTTestLambdaTaskRoleCB13DE78",
          },
          {
            "Ref": "TaskRunnerLambdaFunctionsTaskStatusRole4B498DE5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestLogsGroup54B681C7": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TestPolicyCC05E598": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "cloudwatch:Get*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestPolicyCC05E598",
        "Roles": [
          {
            "Ref": "TaskRunnerLambdaFunctionsLambdaResultsRole1AF5AB18",
          },
          {
            "Ref": "TaskRunnerLambdaFunctionsDLTTestLambdaTaskRoleCB13DE78",
          },
          {
            "Ref": "TaskRunnerLambdaFunctionsLambdaTaskCancelerRoleEEC6795B",
          },
          {
            "Ref": "TaskRunnerLambdaFunctionsTaskStatusRole4B498DE5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestS3Policy438528DE": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Effect": "Deny",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestS3Policy438528DE",
        "Roles": [
          {
            "Ref": "TaskRunnerLambdaFunctionsLambdaResultsRole1AF5AB18",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestTable5769773A": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
